### PR TITLE
add an extra string parameter to terminate the parsing of a file by P…

### PR DESCRIPTION
**Pull request by davydden**
…arameterHandler

### CODE REVIEW

**Comment by bangerth** [comment]
Interesting concept. Out of curiosity, why not pass this string to the various `read_` functions?

**Comment by davydden** [comment]
that would work as well, but I somehow felt that it is move convenient to put it in the constructor, assuming that one normally does not call `read_` more than once for a given parameter file. 

On another hand this piece of information is very much related to the particular input file/stream, so it makes sense to keep the two close.


I am fine either way. I can add an extra default parameter to those two functions:
```
read_input (std::istream &input, const std::string &filename="input file")
read_input (const std::string &filename, const bool optional=false, const bool write_stripped_file=false)
```

**Comment by bangerth** [comment]
I find that the purpose of the string you pass is specific to the process of *reading a file*, and has nothing to do with any of the other functionality of the class. So I think I'd prefer it to be attached to the functions that read text.

You probably also want to consider `read_input_from_string` for completeness.

**Comment by davydden** [comment]
@bangerth i updated the PR.

File: doc/news/changes.h [hunk 1]
```diff
@@ -201,6 +201,13 @@ inconvenience this causes.
 <h3>Specific improvements</h3>
 
 <ol>
+ <li> New: Added an optional string parameter to the ParameterHandler::read_ functions.
```
Comment (bangerth) [inline doc/news/changes.h]: doxygen is going to complain about not being able to find that function. Just list all of them.

File: include/deal.II/base/parameter_handler.h [hunk 2]
```diff
@@ -1576,10 +1576,15 @@ class ParameterHandler : public Subscriptor
    * the file (if that's what the input stream represents) we are reading
    * from; this is only used when creating output for error messages.
    *
+   * If non-empty @p last_line is provided, the ParameterHandler object
+   * will stop parsing lines after encountering @p last_line .
+   * This is handy when adding extra data, that shall be parsed manually.
```
Comment (bangerth) [inline include/deal.II/base/parameter_handler.h]: no comma necessary.

Comment (bangerth) [inline include/deal.II/base/parameter_handler.h]: same below.

**Comment by bangerth** [comment]
Looks good. Please fix the minor text issues, add a test, and merge yourself.

File: doc/news/changes.h [hunk 3]
```diff
@@ -201,6 +201,14 @@ inconvenience this causes.
 <h3>Specific improvements</h3>
 
 <ol>
+ <li> New: Added an optional string parameter to the ParameterHandler::read_input ()
```
Comment (davydden) [inline doc/news/changes.h]: @bangerth is it ok like that or would you prefer to have arguments and two separate `read_input()` functions? I am not sure if doxygen will pick it up from `changes.h`.

Comment (bangerth) [inline doc/news/changes.h]: Yes, this works.

**Pull request merged**